src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

```diff
@@ -906,5 +914,141 @@ private static void ReadToEndOfCurrentObject(JsonReader reader)
 
             throw new JsonSerializationException($"Could not read till end of object, end of stream. Got '{reader.TokenType}'.");
         }
+
+        private static string Intern(string str)
+        {
+            if (str is null)
+            {
+                return null;
+            }
+
+            var interned = string.IsInterned(str);
+            if (interned != null)
+            {
+                return interned;
+            }
+
+            return weakReferenceTable.Intern(str);
+        }
+
+        /// <summary>
+        /// The purpose of this class is to avoid permanently storing duplicate strings which were read in from JSON.
+        /// </summary>
+        private class WeakReferenceTable
+        {
+            public static readonly WeakReferenceTable Instance = new WeakReferenceTable();
+
+            private WeakReferenceTable()
+            {
+            }
+
+            private readonly WeakValueDictionary<string, string> _table = new WeakValueDictionary<string, string>();
+
+            public string Intern(string str)
+            {
+                if (_table.TryGetValue(str, out var value))
+                {
+                    return value;
+                }
+                else
+                {
+                    _table.Add(str, str);
```

Comment:
If key/value is the same, is there a way we can just leverage the keys via a HashSet instead of a Dict? Is this because we want to use a WeakReference value? Won't this double our memory requirements as we're storing a full string copy of the key in the internal `_dict` structure, and then storing an additional weak reference?

I'm sure there's a good reason for all this, just trying to get a better understanding 😄

Reply:
I have a reason, but I wouldn't call it "good" lol. Basically HashSet<T> in netstandard2.0 doesn't have a `Get` method, so I can confirm that an equivalent string is in the table but not retrieve it. Upgrading to 2.1 quickly turned the PR into a huge ordeal, and as you point out HashSet'ing a WeakReference seems fraught.

If y'all have another suggestion given the above info I'm interested because I agree what I've got now isn't ideal.

Reply:
Regardless this is a good reminder to put this info in a comment for future generations.